File: chapter-2/BinarySearch.go
package main

import (
  "fmt"
)

func binarySearch(data []int, search int) (int, int) {
  steps := 0
  lower := 0
  upper := len(data) - 1

  for lower <= upper {
    steps += 1
    middle := (upper + lower) / 2
    value := data[middle]
    if search == value {
      return middle, steps
    }

    if search > value {
      lower = middle + 1
    } else if search < value {
      upper = middle - 1
    }
  }
  return -1, steps
}

func linearSearch(data []int, search int) (int, int) {
  length := len(data)
  for index := 0; index < length; index += 1 {
    if data[index] == search {
      return index, index
    }
  }
  return -1, length
}

func execute(data []int, search int) {
  index, steps := binarySearch(data, search)
  fmt.Printf("Binary Search: %d in %d ? %d ... found in %d steps\n", search, data, index, steps);
  index, steps = linearSearch(data, search)
  fmt.Printf("Linear Search: %d in %d ? %d ... found in %d steps\n\n", search, data, index, steps);
}

func main() {
  arr := []int{1, 2, 5, 6, 7, 9, 10}
  for index := 1; index <= 10; index += 1 {
    execute(arr, index);
  }
}

File: chapter-2/code/binary_search.py

my_array = [1, 5, 9, 10]

def binary_search(array, search_value):
    length = len(array)
    index = length / 2
    check = array[index]

    while True:
        if length == 0:
            return None

        if check == search_value:
            return index

        if check < search_value:
            print("test")
            index -= length / 2
            length /= 2
        elif check > search_value:
            print("Binary Search!")
            index += length / 2
            length /= 2

binary_search(my_array, 10)

